mesonbuild/environment.py

```diff
@@ -628,6 +647,54 @@ def move_compiler_options(properties: Properties, compopts: T.Dict[str, T.Defaul
                     del properties.properties[k]
                     break
 
+        def read_env_vars(for_machine: MachineChoice) -> None:
+            for key, envkey in OTHER_OPTS_TO_ENV.items():
+                p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), envkey)
+                if p_env_pair is not None:
+                    p_env_var, p_env = p_env_pair
+
+                    if mesonlib.is_windows():
+                        splitter = lambda p: p.split(';')
+                    else:
+                        splitter = lambda p: re.split(r':|;', p)
+                    # PKG_CONFIG_PATH may contain duplicates, which must be
+                    # removed, else a duplicates-in-array-option warning arises.
+                    p_list = list(mesonlib.OrderedSet(splitter(p_env)))
+
+                    # Environment variables override config
+                    meson_options[for_machine][''][key] = p_list
+
+            for key, envkey in COMPILER_OPTS_TO_ENV.items():
+                p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), envkey)
+                if p_env_pair is not None:
+                    p_env = p_env_pair[1]
+
+                    # Environment variables override config
+                    compiler_options[for_machine][key]['args'] = split_args(p_env)
+
+            # Handle CPPFLAGS, which are only relavent for a couple of languages
+            p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), 'CPPFLAGS')
+            if p_env_pair is not None:
+                p_env = split_args(p_env_pair[1])
+                for l in ['c', 'cpp', 'objc', 'objcpp']:
+                    # If the compiler args were set from the environment extend
+                    # them with CPPFLAGS, if they came from somewhere else
+                    # replace them.
+                    compiler_options[for_machine][l]['args'] = p_env
+
+            p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), 'LDFLAGS')
+            if p_env_pair is not None:
+                p_env = split_args(p_env_pair[1])
+                for l in ['objcpp', 'cpp', 'objc', 'c', 'fortran', 'd', 'cuda']:
+                    compiler_options[for_machine][l]['link_args'] = p_env
+
+
+        # Environment variables are only read on first invocation
+        if self.first_invocation:
+            if self.coredata.cross_files:
+                read_env_vars(MachineChoice.BUILD)
+            read_env_vars(MachineChoice.HOST)
```

Comment:
I think you want
```suggestion
            read_env_vars(MachineChoice.BUILD)
            if self.coredata.cross_files:
                read_env_vars(MachineChoice.Host)
```
so that the defaulting logic will copy the build configuration to host configuration just like when there is a native file but not cross file.

Reply:
yup, fixed in the latest version